src/lib/transition_handler/validator.ml

```diff
@@ -1,149 +1,122 @@
-                 ~time_received
-             = `Take )
-             "was not better than transition frontier root"
-      with
-      | `Valid ->
-          (* TODO: what conditions should we punish? *)
-          (* TODO:
-      let length = External_transition.protocol_state t |> Protocol_state.blockchain_state |> Blockchain_state.length in
-      log_assert
-        (match Root_history.find (Transition_frontier.root_history frontier) (length - k) with
-        | `Known h -> State_hash.equal h (External_transition.frontier_root_hash t)
-        | `Unknown -> true
-        | `Out_of_bounds ->
-          Logger.info logger "expected root of transition was out of bounds";
-          false)
-        "transition frontier root hash was invalid"
-      *)
-          let%map proof_is_valid =
-            State_proof.verify
-              (External_transition.protocol_state_proof transition)
-              (External_transition.protocol_state transition)
-          in
-          log_assert proof_is_valid "proof was invalid"
-      | x -> return x
-
-  let verify_transition ~staged_ledger ~transition :
-      External_transition.Verified.t Or_error.t Deferred.t =
-    let open Deferred.Or_error.Let_syntax in
-    let diff = External_transition.staged_ledger_diff transition in
-    let%bind verified_diff =
-      Staged_ledger.verified_diff_of_diff staged_ledger diff
-    in
-    return
-      (External_transition.Verified.create
-         ~protocol_state:(External_transition.protocol_state transition)
-         ~protocol_state_proof:
-           (External_transition.protocol_state_proof transition)
-         ~staged_ledger_diff:verified_diff)
+      let rec check_cases = function
+        | [] -> return `Valid
+        | (cond_lazy, failure_reason) :: tail ->
+            if%bind Lazy.force cond_lazy then check_cases tail
+            else return (`Invalid failure_reason)
+      in
+      let validation_cases =
+        [ ( lazy
+              (return
+                 ( `Take
+                 = Consensus.Mechanism.select ~logger
+                     ~existing:(consensus_state root_protocol_state)
+                     ~candidate:(consensus_state protocol_state) ))
+          , "consensus state was not selected over transition frontier root \
+             consensus state" )
+        ; ( lazy
+              (State_proof.verify
+                 (External_transition.protocol_state_proof transition)
+                 protocol_state)
+          , "protocol state proof was not valid" ) ]
+      in
+      let validation_cases =
+        Option.value ~default:validation_cases
+          (Option.map time_received ~f:(fun t ->
+               ( lazy
+                   (return
+                      (Consensus.Mechanism.received_at_valid_time
+                         (consensus_state protocol_state)
+                         ~time_received:t))
+               , "transition was received at an invalid time" )
+               :: validation_cases ))
```

Comment:
This would look cleaner if you use Jane Street's Lazy monad. You can also combine this monad with `Or_error` monad to create lazy short-circuiting. In particular, you could represent validation as `External_transition.Verified.t Or_error.t Deferred.t Lazy.t` or `() Or_error.t Deferred.t Lazy.t`. So, using a monad transformer would help a lot with this.

Reply:
Yeah please try that, it is a bit confusing right now

Reply:
`Or_error.t` doesn't work because I want the caller to explicitly handle the `Invalid` and `Duplicate` cases, which are both errors. I could possibly use a `Result.t`. How would using the deferred lazy monad enable me to clean this up? I'm having a hard time seeing that.

Reply:
Would the scope of  the suggested `Lazy` monad apply to the entire computation here? The laziness given here applies to individual list elements.